draft-trammell-taps-post-sockets.md

```diff
@@ -508,27 +508,43 @@ protocol stack information, and, per {{I-D.pauly-taps-transport-security}},
-due to cost).
-
-These policies are expressed in the Policy Context(s) that are bound to the
-Association.
-Multiple policy contexts can be active at once.
-For example, a system Policy Context can express the administrative preferences
-around network interface and protocol selection, while an application Policy Context
-expresses preferences for use of different transport services.
-Expression of policy contexts and the resolution of conflicts among Policy
-Contexts is currently implementation-specific (the Policy API in the NEAT
-architecture {{NEAT}} provides an example of how this can be done).
+## Configuration {#Configuration}
+
+A Configuration encapsulates an application's preferences around Path
+selection and protocol options.
+
+Each Association has exactly one Configuration, and all Carriers belonging
+to that Association share the same Configuration.
+
+The application cannot modify the Configuration for a Carrier or Association
+once it is set. If a new set of options needs to be used, then the application needs
+a new Carrier or Association instance. This is necessary to ensure that a
+single Carrier can consistently track the Paths and protocol options it uses,
+since it is usually not possible to modify these properties without breaking connectivity.
+
+To influence Path selection, the application can configure a set of requirements,
+preferences, and restrictions concerning which Paths may be selected by the
+Association to use for creating Transients between a Local and a Remote. For example,
+a Configuration can specify that the application prefers Wi-Fi access over
+LTE when roaming on a foreign LTE network, due to monetary cost to the user.
+
+The Association uses the Configuration's Path preferences as a key part of determining
+the Paths to use for its Transients. The Configuration is provided as input when examining
+the complete list of available Paths on the system (to limit the list, or order the Paths by
+preference). The system's policy will further restrict and modify the Path that is ultimately
+selected, using other aspects of the Configuration (protocol options and originating application)
+to select the most appropriate Path.
+
+To influence protocol selection and options, the Configuration contains one
+or more allowed Protocol Stack Configurations. Each of these is comprised of
+application- and transport-layer protocols that may be used together to communicate
+to the Remote, along with any protocol-specific options. For example, a
+Configuration could specify two alternate, but equivalent, protocol stacks:
+one using HTTP/2 over TLS over TCP, and the other using QUIC over UDP.
+Alternatively, the Configuration could specify two protocol stacks with the same
+protocols, but different protocol options: one using TLS with TLS 1.3 0-RTT enabled
+and TCP with TCP Fast-Open enabled, and one using TLS with out 0-RTT and TCP
+without TCP Fast-Open.
 
```

Comment:
Does the Protocol Stack Configuration need to be explicitly separate from the rest of the Configuration, or can we just note that the requirements expressed in the Configuration will be used to select Paths as well as to choose among possible PSIs?

Reply:
The reason I separated the Protocol Stack Configuration out as a sub-object is that the Configuration can hold an array of PSCs, to describe self-coherent options that may be raced, or we can fall back between.

I'm doing this to lean towards applications specifying which protocols they allow—I know that the NEAT side of TAPS stuff may want to just infer that you can try SCTP, etc, whenever, but I think there needs to be some indication from the application that there is a set of known supported configurations before we go and try random things because certain protocols have properties that match the desires of the applications.